File: Park/CodeTree/연속적이지만 직전 상황에 영향을 받는 DP/계단 수.py
# 숫자의 길이 n을 입력받습니다.
n = int(input())

# dp 배열 초기화
dp = [[0] * 10 for _ in range(n + 1)]

# 첫 번째 자리에 올 수 있는 숫자들의 경우의 수를 초기화합니다.
for i in range(1, 10):
    dp[1][i] = 1

MOD = 1000000007

# 동적 프로그래밍을 사용하여 각 자리수마다 가능한 숫자들의 합을 계산합니다.
for i in range(1, n):
    for j in range(10):
        if j > 0:
            dp[i + 1][j - 1] = (dp[i + 1][j - 1] + dp[i][j]) % MOD
        if j < 9:
            dp[i + 1][j + 1] = (dp[i + 1][j + 1] + dp[i][j]) % MOD

# 최종 결과를 계산합니다.
ans = sum(dp[n]) % MOD

# 계산된 결과를 출력합니다.
print(ans)

File: jhyun/DP2/연속적이지만 직전 상황에 영향을 받는 DP/계단_수.java
import java.util.*;
import java.io.*;

public class Main {
    static final int MOD = 1000000007;

    static int n;
    static int[][] dp = new int[1005][15]; //i는 자릿수, j는 해당 자릿수에서의 숫자

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());

        for(int i = 1; i <= 9; i++){
            dp[1][i] = 1;
        }

        // 동적 프로그래밍을 사용하여 각 자리수마다 가능한 숫자들의 합을 계산
        // dp[i][j] :: i자릿수 까지 봤을 때, 마지막 숫자가 j인 가짓수
        for(int i = 1; i < n; i++){
            for(int j = 0; j <= 9; j++){
                //숫자가 감소하는 경우
                if(j > 0){
                    dp[i+1][j-1] = (dp[i+1][j-1] + dp[i][j]) % MOD;
                }
                //숫자가 증가하는 경우
                if(j < 9){
                    dp[i+1][j+1] = (dp[i+1][j+1] + dp[i][j]) % MOD;
                }
            }
        }

        int ans = 0;
        for(int j = 0; j <= 9; j++){
            ans = (ans + dp[n][j]) % MOD;
        }

        System.out.println(ans);
    }
}